File: hooks/shell-guard.py
#!/usr/bin/env python3
"""beforeShellExecution hook: warn on risky shell commands.

Reads the proposed shell command, scans for risky and severe patterns, and
emits a structured JSON decision.

Pattern policy stays narrow on purpose so the hook does not turn into a
discretionary gate for ordinary work:

- **warning** patterns (force-push, --no-verify, rm -rf, hard reset, branch -D,
  checkout discard) keep `permission=allow`; the hook just surfaces a reminder
  through `user_facing_message`.
- **severe** patterns are tightly bounded to operations that would corrupt
  repo-owned state files (`.cursor/state/workflow-state*.json`) or the local
  plugin install path (`~/.cursor/plugins/local/oh-my-cursor`). These set
  `permission=ask` so Cursor can confirm with the user before executing.

The hook itself never deletes anything, never rewrites the command, and always
exits 0. Cursor consumes the `permission` field, not the exit code.
"""
from __future__ import annotations

import json
import re
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))
from _trace import trace as _trace  # noqa: E402


COMMAND_FIELDS = ("command", "commandLine", "command_line", "shell", "cmd", "input")

WARNING_PATTERNS = {
    "force-push": re.compile(r"\bgit\s+push\b[^|;&]*--force(?:-with-lease)?\b", re.IGNORECASE),
    "skip-hooks": re.compile(r"--no-verify\b", re.IGNORECASE),
    "skip-gpg-sign": re.compile(r"--no-gpg-sign\b", re.IGNORECASE),
    "rm-rf-broad": re.compile(r"\brm\s+-[a-zA-Z]*r[a-zA-Z]*f[a-zA-Z]*\b", re.IGNORECASE),
    "git-reset-hard": re.compile(r"\bgit\s+reset\s+--hard\b", re.IGNORECASE),
    "git-clean-force": re.compile(r"\bgit\s+clean\b[^|;&]*-[A-Za-z]*f[A-Za-z]*\b", re.IGNORECASE),
    "checkout-discard": re.compile(r"\bgit\s+checkout\s+--\s+\.", re.IGNORECASE),
    "branch-delete-force": re.compile(r"\bgit\s+branch\b[^|;&]*-D\b", re.IGNORECASE),
}

SEVERE_PATTERNS = {
    "destroy-state-file": re.compile(
        r"\b(?:rm|truncate|sed|awk|tee)\b[^|;&]*\.cursor/state/workflow-state(?:\.schema|\.example)?\.json\b",
        re.IGNORECASE,
    ),
    "redirect-overwrite-state-file": re.compile(
        r">\s*\"?\.cursor/state/workflow-state(?:\.schema|\.example)?\.json\b",
        re.IGNORECASE,
    ),
    "destroy-local-plugin-install": re.compile(
        r"\brm\s+-[a-zA-Z]*r[a-zA-Z]*f[a-zA-Z]*\b[^|;&]*~?/?\.cursor/plugins/local/oh-my-cursor\b",
        re.IGNORECASE,
    ),
    "destroy-local-plugin-root": re.compile(
        r"\brm\s+-[a-zA-Z]*r[a-zA-Z]*f[a-zA-Z]*\b[^|;&]*~?/?\.cursor/plugins/local/?\s*$",
        re.IGNORECASE,
    ),
}


def _read_payload() -> dict:
    raw = sys.stdin.read()
    if not raw.strip():
        return {}
    try:
        value = json.loads(raw)
    except json.JSONDecodeError:
        return {"_invalid_json": True}
    return value if isinstance(value, dict) else {"payload": value}


def _extract_command(payload: dict) -> str:
    for key in COMMAND_FIELDS:
        value = payload.get(key)
        if isinstance(value, str) and value.strip():
            return value
    for nested_key in ("tool", "arguments", "tool_call", "toolCall"):
        nested = payload.get(nested_key) if isinstance(payload, dict) else None
        if isinstance(nested, dict):
            for key in COMMAND_FIELDS:
                value = nested.get(key)
                if isinstance(value, str) and value.strip():
                    return value
    return ""


def _scan(command: str, patterns: dict[str, re.Pattern[str]]) -> list[str]:
    matched: list[str] = []
    for label, pattern in patterns.items():
        if pattern.search(command):
            matched.append(label)
    return matched


def main() -> int:
    payload = _read_payload()
    if payload.get("_invalid_json"):
        print(json.dumps({
            "status": "pass",
            "fail_open": True,
            "permission": "allow",
            "user_facing_message": "Shell-guard input was not JSON; skipped.",
        }))
        return 0

    command = _extract_command(payload)
    warnings = _scan(command, WARNING_PATTERNS)
    severe = _scan(command, SEVERE_PATTERNS)

    if severe:
        permission = "ask"
        message = (
            "Shell-guard flagged a severe pattern that would corrupt repo-owned state or the local plugin install: "
            + ", ".join(severe)
            + ". Use scripts/workflow-state.py for state edits and scripts/install-local-plugin.sh for the local plugin path."
        )
        status = "severe"
    elif warnings:
        permission = "allow"
        message = (
            "Shell-guard noticed risky patterns: "
            + ", ".join(warnings)
            + ". Confirm the command is intentional before continuing."
        )
        status = "warning"
    else:
        permission = "allow"
        message = "Shell-guard saw no risky patterns."
        status = "pass"

    output = {
        "status": status,
        "fail_open": not severe,
        "permission": permission,
        "user_facing_message": message,
        "command_excerpt": command[:200],
        "warnings": warnings,
        "severe": severe,
    }
    _trace({
        "hook": "shell-guard",
        "event": "beforeShellExecution",
        "status": status,
        "permission": permission,
        "command_excerpt": command[:120],
        "warnings": warnings,
        "severe": severe,
    })
    print(json.dumps(output, ensure_ascii=False))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: hooks/shell-guard.ts
import * as fs from 'node:fs';
import * as process from 'node:process';
import { trace } from './_trace.ts';

const COMMAND_FIELDS = ['command', 'commandLine', 'command_line', 'shell', 'cmd', 'input'];

const WARNING_PATTERNS: Record<string, RegExp> = {
  'force-push': /git\s+push\b[^|;&]*--force(?:-with-lease)?\b/i,
  'skip-hooks': /--no-verify\b/i,
  'skip-gpg-sign': /--no-gpg-sign\b/i,
  'rm-rf-broad': /rm\s+-[a-zA-Z]*r[a-zA-Z]*f[a-zA-Z]*\b/i,
  'git-reset-hard': /git\s+reset\s+--hard\b/i,
  'git-clean-force': /git\s+clean\b[^|;&]*-[A-Za-z]*f[A-Za-z]*\b/i,
  'checkout-discard': /git\s+checkout\s+--\s+\./i,
  'branch-delete-force': /git\s+branch\b[^|;&]*-D\b/i,
};

const SEVERE_PATTERNS: Record<string, RegExp> = {
  'destroy-state-file': /(?:rm|truncate|sed|awk|tee)\b[^|;&]*\.cursor\/state\/workflow-state(?:\.schema|\.example)?\.json\b/i,
  'redirect-overwrite-state-file': />\s*\"?\.cursor\/state\/workflow-state(?:\.schema|\.example)?\.json\b/i,
  'destroy-local-plugin-install': /rm\s+-[a-zA-Z]*r[a-zA-Z]*f[a-zA-Z]*\b[^|;&]*~?\/?\.cursor\/plugins\/local\/oh-my-cursor\b/i,
  'destroy-local-plugin-root': /rm\s+-[a-zA-Z]*r[a-zA-Z]*f[a-zA-Z]*\b[^|;&]*~?\/?\.cursor\/plugins\/local\/?\s*$/i,
};

function readPayload(): any {
  try {
    const raw = fs.readFileSync(0, 'utf-8');
    if (!raw.trim()) {
      return {};
    }
    return JSON.parse(raw);
  } catch (err: any) {
    return { _invalid_json: true };
  }
}

function extractCommand(payload: any): string {
  if (!payload || typeof payload !== 'object') {
    return '';
  }
  for (const key of COMMAND_FIELDS) {
    const value = payload[key];
    if (typeof value === 'string' && value.trim()) {
      return value;
    }
  }
  for (const nestedKey of ['tool', 'arguments', 'tool_call', 'toolCall']) {
    const nested = payload[nestedKey];
    if (nested && typeof nested === 'object' && !Array.isArray(nested)) {
      for (const key of COMMAND_FIELDS) {
        const value = nested[key];
        if (typeof value === 'string' && value.trim()) {
          return value;
        }
      }
    }
  }
  return '';
}

function scan(command: string, patterns: Record<string, RegExp>): string[] {
  const matched: string[] = [];
  for (const [label, regex] of Object.entries(patterns)) {
    if (regex.test(command)) {
      matched.push(label);
    }
  }
  return matched;
}

function main(): number {
  const payload = readPayload();
  if (payload?._invalid_json) {
    console.log(JSON.stringify({
      status: 'pass',
      fail_open: true,
      permission: 'allow',
      user_facing_message: 'Shell-guard input was not JSON; skipped.',
    }));
    return 0;
  }

  const command = extractCommand(payload);
  const warnings = scan(command, WARNING_PATTERNS);
  const severe = scan(command, SEVERE_PATTERNS);

  let permission = 'allow';
  let message = 'Shell-guard saw no risky patterns.';
  let status = 'pass';

  if (severe.length > 0) {
    permission = 'ask';
    message = (
      'Shell-guard flagged a severe pattern that would corrupt repo-owned state or the local plugin install: ' +
      severe.join(', ') +
      '. Use scripts/workflow-state.py for state edits and scripts/install-local-plugin.ts for the local plugin path.'
    );
    status = 'severe';
  } else if (warnings.length > 0) {
    permission = 'allow';
    message = (
      'Shell-guard noticed risky patterns: ' +
      warnings.join(', ') +
      '. Confirm the command is intentional before continuing.'
    );
    status = 'warning';
  }

  const output = {
    status,
    fail_open: severe.length === 0,
    permission,
    user_facing_message: message,
    command_excerpt: command.slice(0, 200),
    warnings,
    severe,
  };

  trace({
    hook: 'shell-guard',
    event: 'beforeShellExecution',
    status,
    permission,
    command_excerpt: command.slice(0, 120),
    warnings,
    severe,
  });

  console.log(JSON.stringify(output));
  return 0;
}

process.exit(main());
